tower-batch/src/worker.rs

```diff
@@ -43,66 +44,125 @@ pub(crate) struct Handle {
+                    self.failed(e.into());
+                    let _ = tx.send(Err(self
+                        .failed
+                        .as_ref()
+                        .expect("Worker::failed did not set self.failed?")
+                        .clone()));
+                }
             }
+        }
+    }
 
-            tracing::trace!("dropping cancelled buffered request");
+    async fn flush_service(&mut self) {
+        if let Err(e) = self
+            .service
+            .ready_and()
+            .and_then(|svc| svc.call(BatchControl::Flush))
+            .await
+        {
+            self.failed(e.into());
         }
+    }
 
-        // Get the next request
-        while let Some(mut msg) = ready!(Pin::new(&mut self.rx).poll_recv(cx)) {
-            if msg.tx.poll_closed(cx).is_pending() {
-                tracing::trace!("processing new request");
-                return Poll::Ready(Some((msg, true)));
+    pub async fn run(mut self) {
+        use futures::future::Either::{Left, Right};
+        // The timer is started when the first entry of a new batch is
+        // submitted, so that the batch latency of all entries is at most
+        // self.max_latency. However, we don't keep the timer running unless
+        // there is a pending request to prevent wakeups on idle services.
+        let mut timer: Option<Delay> = None;
+        let mut pending_items = 0usize;
+        loop {
+            match timer {
+                None => match self.rx.next().await {
+                    // The first message in a new batch.
+                    Some(msg) => {
+                        let span = msg.span;
+                        self.process_req(msg.request, msg.tx)
+                            // Apply the provided span to request processing
+                            .instrument(span)
+                            .await;
+                        timer = Some(delay_for(self.max_latency));
+                        pending_items = 1;
+                    }
+                    // No more messages, ever.
+                    None => return,
+                },
+                Some(delay) => {
+                    // Wait on either a new message or the batch timer.
+                    match futures::future::select(self.rx.next(), delay).await {
+                        Left((Some(msg), delay)) => {
+                            let span = msg.span;
+                            self.process_req(msg.request, msg.tx)
+                                // Apply the provided span to request processing.
+                                .instrument(span)
+                                .await;
+                            pending_items += 1;
+                            // Check whether we have too many pending items.
+                            if pending_items >= self.max_items {
+                                // XXX(hdevalence): what span should instrument this?
+                                self.flush_service().await;
```

Comment:
This should be instrumented with some span, but I'm not sure what the "correct" context is here.  This work doesn't happen in the context of a particular request, so it shouldn't be instrumented as such -- should the worker have its own span?

Reply:
I'd say that or maybe try to grab a span when the Batch layer is constructed and report that location for the worker future